dockerfiles/Dockerfile

```diff
@@ -0,0 +1,170 @@
+# syntax=docker/dockerfile:1.3-labs
+ARG OPENMS_VERSION=3.1.0
+ARG OPENMS_LIBRARY_BUILD_DIR="/openms-build"
+ARG OPENMS_LIBRARY_DIR="/OpenMS"
+ARG CONTRIB_TAG=latest
+ARG CONTRIB_BASE_IMAGE=ghcr.io/openms/contrib:${CONTRIB_TAG}
+ARG CONTRIB_BUILD_DIR="/contrib-build"
+ARG THIRDPARTY_BRANCH="master"
+ARG THIRDPARTY_DIR="/thirdparty"
+ARG NUM_BUILD_CORES=8
+ARG MAKEFLAGS="-j${NUM_BUILD_CORES}"
+
+
+### OpenMS base ###
+FROM ${CONTRIB_BASE_IMAGE} as base
+ARG OPENMS_VERSION
+
+# the main reason we're pulling this in is to make sure that the metadata
+# below is consistently formatted for releases
+LABEL version="${OPENMS_VERSION}"
+LABEL software.version="${OPENMS_VERSION}-Ubuntu22.04"
+LABEL org.opencontainers.image.source https://github.com/OpenMS/OpenMS
+
+
+### OpenMS library ###
+FROM base as library
+ARG CONTRIB_BUILD_DIR
+ARG OPENMS_LIBRARY_DIR
+ARG OPENMS_LIBRARY_BUILD_DIR
+ARG MAKEFLAGS
+ENV MAKEFLAGS="${MAKEFLAGS}"
+
+# NOTE: this assumes that the context is root of the OpenMS repo
+COPY . ${OPENMS_LIBRARY_DIR}
+WORKDIR ${OPENMS_LIBRARY_BUILD_DIR}
+RUN /bin/bash -c "cmake -DCMAKE_BUILD_TYPE='Release' -DCMAKE_PREFIX_PATH='${CONTRIB_BUILD_DIR}/;/usr/;/usr/local' -DBOOST_USE_STATIC=OFF ${OPENMS_LIBRARY_DIR}"
+RUN make OpenMS
+LABEL software="OpenMS (library)"
+
+
+### OpenMS tools build ###
+FROM library as tools-build
+ARG OPENMS_LIBRARY_DIR
+ARG OPENMS_LIBRARY_BUILD_DIR
+ARG MAKEFLAGS
+ARG THIRDPARTY_BRANCH
+ARG THIRDPARTY_DIR
+ENV MAKEFLAGS="${MAKEFLAGS}"
+
+RUN apt-get update && apt-get install -y --no-install-recommends --no-install-suggests openjdk-17-jdk
+
+WORKDIR ${OPENMS_LIBRARY_DIR}
+RUN <<-EOF
+    mkdir /thirdparty
+    # note, we are not using the submodule here because the .git folder
+    # is not copied to this stage. Instead, we clone the repo directly.
+    git clone --branch=${THIRDPARTY_BRANCH} --depth=1 https://github.com/OpenMS/THIRDPARTY.git
+
+    # copying only the relevant binaries and leaving the rest behind
+    cp -r THIRDPARTY/All/* ${THIRDPARTY_DIR}
+    cp -r THIRDPARTY/Linux/64bit/* ${THIRDPARTY_DIR}
+
+    rm -rf THIRDPARTY
+EOF
+
+ENV PATH="${THIRDPARTY_DIR}/LuciPHOr2:${THIRDPARTY_DIR}/MSGFPlus:${THIRDPARTY_DIR}/Sirius:${THIRDPARTY_DIR}/ThermoRawFileParser:${THIRDPARTY_DIR}/Comet:${THIRDPARTY_DIR}/Fido:${THIRDPARTY_DIR}/MaRaCluster:${THIRDPARTY_DIR}/Percolator:${THIRDPARTY_DIR}/SpectraST:${THIRDPARTY_DIR}/XTandem:${THIRDPARTY_DIR}/Sage:${PATH}"
+WORKDIR ${OPENMS_LIBRARY_BUILD_DIR}
+# RUN make TOPP && rm -rf src doc CMakeFiles
+RUN make TOPP
+WORKDIR /
+ENV PATH="${OPENMS_LIBRARY_BUILD_DIR}/bin:${PATH}"
+
+### OpenMS tools runtime ###
+FROM ubuntu:22.04 AS tools
+ARG OPENMS_LIBRARY_DIR
+ARG OPENMS_LIBRARY_BUILD_DIR
+ARG CONTRIB_BUILD_DIR
+ARG OPENMS_LIBRARY_DIR
+ARG THIRDPARTY_DIR
+
+RUN apt-get update \
+  && apt-get install -y --no-install-recommends --no-install-suggests \
+    libqt5opengl5 \
```

Comment:
We might think about a distinction between tools and gui-tools.
GUI tools will install a ton of other stuff.

Reply:
Good point. Is it just a matter of excluding dependencies for GUI tools in the `tools` and including them in `gui-tools`? Or are there some GUI only tools that we shouldn't build at all?

Reply:
You can re-use an existing OpenMS build with "-DWITH_GUI=OFF", install additional dependencies such as qt5 gui components and then re-configure CMake with "-DWITH_GUI=ON". Then one can use `make OpenMS_GUI` to build the GUI tools only.
This could be just an additional layer on top of the tools image.

Reply:
Alright, so I did this and the difference is 7.7 MB for the library image and 41 MB for the tools image (they're just under 900 MB). The `Dockerfile` is quite a bit more complex though. I'm not sure it's worth the tradeoff especially when we add a `-thirdparty` variant which will make a much bigger difference.

It's possible I'm doing something wrong though. I did as you said:
- for the `library-build` stage I configure with `-DWITH_GUI=OFF`
- for the `library-build-gui` stage reconfigured with  `-DWITH_GUI=ON`, then `make OpenMS_GUI`
- I only install `libqt5opengl5` in the `library` stage
- the `tools` and `tools-gui` images are mostly the same -- I still `make TOPP` in the build stage. Then, I copy the library from the corresponding build stage, and I install `openjdk-17-jre` for the GUI and `openjdk-17-jre-headless` for the regular image.

IMO, unless I've missed something I'd rather aim for a `-thirdparty` variant, merge, and then revisit the GUI issue.